src/lib.rs

```diff
@@ -79,7 +79,7 @@ impl Connection {
         let _ = self.sender.send(resp.into());
         match &self.receiver.recv_timeout(std::time::Duration::from_secs(30)) {
             Ok(Message::Notification(n)) if n.is_exit() => (),
-            m => Err(ProtocolError(format!("unexpected message during shutdown: {:?}", m)))?,
+            m => return Err(ProtocolError(format!("unexpected message during shutdown: {:?}", m))),
```

Comment:
I personally prefer the `Err(...)?` idiom, but, given that I fail to use it consistently here, I am fine with this change!